Announcements: Add indices to Discussion table

bleistivt opened the pull request:
This should mainly help small setups without an active cache.
I tested this on a board with 7k discussion and around a dozen announcements. Announcement queries that used to take ~5ms are now  0,27ms. The single index for `Anounce` was required to make the /discussions query fast as well (and not just single categories).

fixes #3566

alex-mtl commented:
@bleistivt Could you please provide more details about this issue. What are user story behind? What page, ajax call, api doyou want to fix/ improve?
What is the current query? Any reference to the source code?

bleistivt commented:
@alex-mtl It's pretty much already summed up in the issue i linked above (#3566) but as for actual queries:

https://github.com/vanilla/vanilla/blob/5df9418c71a0434c8364a270a8a76e3879a7bdf2/applications/vanilla/models/class.discussionmodel.php#L1146

This announcements query is slow, both for `/discussions` where it looks like this:

    ...
    where d.Announce = 1

...and single categories `/categories/$categoryID` where it looks like this:

    ...
    where d.Announce > 0
    and d.CategoryID = $categoryID

Now as to why 2 indexes are needed:

What @DaazKu suggested [here](https://github.com/vanilla/vanilla/issues/3566#issuecomment-190908965) was actually:

`CREATE INDEX IX_Discussion_AnnounceCategory ON GDN_Discussion (‘Announce’, ‘CategoryID’)`

However, there is a fundamental design limitation of `Gdn_MySQLStructure` https://github.com/vanilla/vanilla/blob/818e6a6dc387dd47a0a9fa78c6899aebf6ba7ecb/library/database/class.mysqlstructure.php#L365

It only supports indexes in the order that the columns are defined. This would be:

`CREATE INDEX IX_Discussion_AnnounceCategory ON GDN_Discussion (‘CategoryID’, ‘Announce’)`

Since only the first column in a multi-index helps MySQL do a range search I added another single index for `Announce` in my original PR and observed a speedup. But I got curious, so I just did a simple test on an actual forum (mean of 5 requests):

    no index
    4028μs /discussions (3700, 4010, 4060, 4520, 3850)
    3892μs /categories/$categoryID (4080, 3830, 3840, 3830, 3880)

    index (Announce)
     163μs /discussions (162, 171, 161, 159, 164)
     167μs /categories/$categoryID (173, 178, 169, 160, 155)

    index (CategoryID, Announce)
    4026μs /discussions (3920, 4580, 3840, 3970, 3820)
    4788μs /categories/$categoryID (4030, 5970, 3860, 4390, 5690)

    index (Announce, CategoryID) - not supported
     180μs /discussions (167, 156, 213, 173, 192)
     188μs /categories/$categoryID (162, 253, 168, 180, 179)

    index (Announce) && index (CategoryID, Announce) - this PR
     220μs /discussions (176, 267, 186, 218, 255)
     190μs /categories/$categoryID (159, 233, 175, 192, 191)

Which shows me the combined index is pretty much useless because of the way `Gdn_MySQLStructure` handles indeces.
I modified the pull request to reflect my findings.
